hbase-backup/src/main/java/org/apache/hadoop/hbase/backup/replication/ContinuousBackupStagingManager.java

```diff
@@ -0,0 +1,504 @@
+        Utils.logPeerId(continuousBackupManager.getPeerId()));
+      flushWalFiles();
+      backupWalFiles();
+    } catch (IOException e) {
+      LOG.error("{} Error during periodic WAL flush: {}",
+        Utils.logPeerId(continuousBackupManager.getPeerId()), e.getMessage(), e);
+    }
+  }
+
+  private void flushWalFiles() {
+    lock.lock();
+    try {
+      for (Map.Entry<Path, ContinuousBackupWalWriter> entry : walWriterMap.entrySet()) {
+        flushWalData(entry.getKey(), entry.getValue());
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void flushWalData(Path walDir, ContinuousBackupWalWriter writer) {
+    if (writer.hasAnyEntry()) {
+      LOG.debug("{} Flushing WAL data for {}", Utils.logPeerId(continuousBackupManager.getPeerId()),
+        walDir);
+      closeWriter(writer);
+      walWriterMap.put(walDir, createNewContinuousBackupWalWriter(walDir));
+    } else {
+      LOG.debug("{} No WAL data to flush for {}",
+        Utils.logPeerId(continuousBackupManager.getPeerId()), walDir);
+    }
+  }
+
+  private void closeWriter(ContinuousBackupWalWriter writer) {
+    try {
+      writer.close();
+    } catch (IOException e) {
+      LOG.error("{} Error occurred while closing WAL writer: ",
+        Utils.logPeerId(continuousBackupManager.getPeerId()), e);
+    }
+  }
+
+  private ContinuousBackupWalWriter createNewContinuousBackupWalWriter(Path walDir) {
+    try {
+      return new ContinuousBackupWalWriter(walStagingFs, walStagingDir, walDir, conf);
+    } catch (IOException e) {
+      throw new UncheckedIOException("Failed to create WAL Writer for " + walDir, e);
+    }
+  }
+
+  private void startBackupExecutor() {
+    int threads =
+      conf.getInt(CONF_STATED_FILES_BACKUP_THREADS, DEFAULT_STATED_FILES_BACKUP_THREADS);
+    backupExecutor = Executors.newFixedThreadPool(threads);
+  }
+
+  /**
+   * Stages WAL entries and bulk-load files for a specific table. Ensures that WAL entries are
+   * written and bulk-loaded files are registered to prevent deletion by the HFileCleaner thread.
+   * @param tableName     the name of the table
+   * @param walEntries    the list of WAL entries to stage
+   * @param bulkLoadFiles the list of bulk-load files to stage
+   * @throws IOException if there is an error staging the entries or files
+   */
+  public void stageEntries(TableName tableName, List<WAL.Entry> walEntries,
+    List<Path> bulkLoadFiles) throws IOException {
+    lock.lock();
+    try {
+      String namespace = tableName.getNamespaceAsString();
+      String table = tableName.getQualifierAsString();
+
+      Path walDir = WALUtils.getWalDir(namespace, table);
+      ContinuousBackupWalWriter continuousBackupWalWriter = getContinuousBackupWalWriter(walDir);
+
+      continuousBackupWalWriter.write(walEntries, bulkLoadFiles);
+
+      // prevent bulk-loaded files from deleting HFileCleaner thread
+      stagedBulkloadFileRegistry.addStagedFiles(bulkLoadFiles);
```

Comment:
We might have a race here:
1. `bulkLoadFiles` are written by `WalWriter`
2. `bulkLoadFiles` gets added to the exception list for `HFileCleaner`

If `HFileCleaner` runs between these 2 events, it will delete files which recently have been staged. I'm not sure yet how to do this properly.

Reply:
No, that scenario doesn't occur. The replication framework prevents the deletion of files using `ReplicationHFileCleaner`. Files are only deleted once the WALs and bulkload files are successfully replicated, as indicated by a success return from the replicate() method.

However, in our case, the bulkloaded files are still in the staging area and haven’t been copied yet. Therefore, we had to implement an additional cleaner for this purpose.